cudax/include/cuda/experimental/__async/sender/completion_signatures.cuh

```diff
@@ -24,24 +24,418 @@
+_CUDAX_API constexpr auto
+completion_signatures<_Sigs...>::transform_reduce(_Transform __transform, _Reduce __reduce) const
+  -> _CUDA_VSTD::__call_result_t<_Reduce, _CUDA_VSTD::__call_result_t<_Transform, _Sigs*>...>
 {
-  struct __partitioned;
+  return __reduce(__transform(static_cast<_Sigs*>(nullptr))...);
+}
+
+template <class _Ty>
+_CCCL_CONCEPT __valid_completion_signatures = detail::__is_specialization_of<_Ty, completion_signatures>;
+
+template <class _Derived>
+struct _CCCL_TYPE_VISIBILITY_DEFAULT __compile_time_error // : ::std::exception
+{
+  __compile_time_error() = default;
+
+  const char* what() const noexcept // override
+  {
+    return _CCCL_TYPEID(_Derived*).name();
+  }
 };
 
-template <class _CompletionSignatures>
-using __partitioned_completions_of = typename _CompletionSignatures::__partitioned;
+template <class _Data, class... _What>
+struct _CCCL_TYPE_VISIBILITY_DEFAULT __sender_type_check_failure
+    : __compile_time_error<__sender_type_check_failure<_Data, _What...>>
+{
+  __sender_type_check_failure() = default;
+
+  explicit constexpr __sender_type_check_failure(_Data data)
+      : data_(data)
+  {}
+
+  _Data data_{};
+};
+
+struct _CCCL_TYPE_VISIBILITY_DEFAULT dependent_sender_error // : ::std::exception
+{
+  _CUDAX_TRIVIAL_API char const* what() const noexcept // override
+  {
+    return what_;
+  }
+
+  char const* what_;
+};
+
+template <class _Sndr>
+struct _CCCL_TYPE_VISIBILITY_DEFAULT __dependent_sender_error : dependent_sender_error
+{
+  _CUDAX_TRIVIAL_API constexpr __dependent_sender_error() noexcept
+      : dependent_sender_error{"This sender needs to know its execution " //
+                               "environment before it can know how it will complete."}
+  {}
+
+  _CCCL_HOST_DEVICE __dependent_sender_error operator+();
+
+  template <class _Ty>
+  _CCCL_HOST_DEVICE __dependent_sender_error& operator,(_Ty&);
+
+  template <class... _What>
+  _CCCL_HOST_DEVICE _ERROR<_What...>&operator,(_ERROR<_What...>&);
+};
+
+#if defined(__cpp_constexpr_exceptions) // C++26, https://wg21.link/p3068
+#  define _CUDAX_LET_COMPLETIONS(...)                  \
+    if constexpr ([[maybe_unused]] __VA_ARGS__; false) \
+    {                                                  \
+    }                                                  \
+    else
+
+template <class... What, class... Values>
+[[noreturn, nodiscard]] _CUDAX_API consteval completion_signatures<> invalid_completion_signature(Values... values)
+{
+  if constexpr (sizeof...(Values) == 1)
+  {
+    throw __sender_type_check_failure<Values..., What...>(values...);
+  }
+  else
+  {
+    throw __sender_type_check_failure<__tuple<Values...>, What...>(__tupl{values...});
+  }
+}
+
+template <class... _Sndr>
+[[noreturn, nodiscard]] _CUDAX_API consteval auto __dependent_sender() -> completion_signatures<>
+{
+  throw __dependent_sender_error<_Sndr...>();
+}
+#else
+
+#  define _CUDAX_PP_EAT_AUTO_auto(_ID)    _ID _CCCL_PP_EAT _CCCL_PP_LPAREN
```

Comment:
Should we undef those macros after use?

Reply:
unfortunately we can't since they are used by `_CUDAX_LET_COMPLETIONS`, which gets used in other headers.